### ISSUE TITLE
Reddit's API changes

### ISSUE THREAD

**Issue by rubenelshof** [spikecodes/libreddit#785]
Yesterday Reddit announced changes to it's API.
It includes a paid option for third parties who require additional capabilities, higher usage limits, and broader usage rights.
Further it limits access to mature content & rate limiting will be enforced.
These changes will be effective on June 19, 2023.
See the post below
https://www.reddit.com/r/reddit/comments/12qwagm/an_update_regarding_reddits_api/

I would like to know how these changes might affect Libreddit.

**Comment by sigaloid** [comment]
An important question we need to consider is what we do if they lock down unauthenticated requests. Is webpage scraping something that we should implement? I think it's really important to have something like Libreddit but it's a different ballgame if it's a cat and mouse game rather than just piggy-backing on the API routes which is (assumed to be) allowed to be used. 

[Web scraping _is_ legal, technically, in the USA.](https://techcrunch.com/2022/04/18/web-scraping-legal-court/) But it might not be good to stoop into the grey area especially in other countries. 

We'll have to discuss this if it happens, but it's good to ask early so we can think about it.

**Comment by r7l** [comment]
Without any 3rd party apps, it will be a mess using Reddit. The site is clunky already and will only become worse in future seeing the road they are on already. Not sure if it's actually such a big deal to go the [Nitter](https://github.com/zedeus/nitter) way. Same as Libreddit they don't have login / account support and aren't using Twitter API.

**Comment by sigaloid** [comment]
It might not be terrible as an alternative, but it would likely require abandoning much of the codebase and probably starting from scratch, besides the front end templates. We could also run into rate limits, anti-DDoS, bot detection, etc. Just something to consider. I'd be willing to help with that transition if it becomes necessary.

**Comment by nickelswitte** [comment]
Hello,
I want to express my appreciation for this project. I am using reddit on a regular basis to find information or answers to many of my problems and really dislike using their website. LibReddit offered me a great, simple and safe way to cut away all that trash and just the see content. I hope this project will survive this API policy change and very interested in keep using it. Thank you for the work 🚀

**Comment by sphars** [comment]
The developer of the well-known iOS app, Apollo, had a call with reddit themselves to discuss this. Might have some good info here with regard to API access come the change: https://www.reddit.com/r/apolloapp/comments/12ram0f/had_a_few_calls_with_reddit_today_about_the/

**Comment by sigaloid** [comment]
Assuming they leave those endpoints up, yes, that's something we could do. It would be a downgrade from the info we get from the API, but it's not a dealbreaker.

**Comment by rubenelshof** [comment]
@mbestavros Your link goes back to this issue.
The correct link.
https://www.reddit.com/r/apolloapp/comments/13ws4w3/had_a_call_with_reddit_to_discuss_pricing_bad/

**Comment by sigaloid** [comment]
Yep, very very troubling discovery here with the pricing... this would likely price out everyone, even in the case of a Libreddit operator wanting to run a small instance and is willing to cover $5-10 in API fees. It's hard to tell how much exactly it would cost. 

I'm considering adding a stats counter that counts purely the number of requests to Reddit in order to provide an estimate to this sort of thing, but this kind of analytics can have privacy downfalls. I discussed this half a year ago in https://github.com/libreddit/libreddit/issues/624#issuecomment-1305115792 and #685. Going to ask the community about that possibility in https://github.com/libreddit/libreddit/discussions/807.

In any case the pricing given is pretty bonkers. :(

**Comment by sigaloid** [comment]
> The API (auth and anonymous) will likely still exist, because Reddit has to serve the official mobile client (logged in and guest users).

The Reddit app doesn't use the same endpoints as we do. It uses a private API that we can't simply reverse engineer. It's entirely possible (and I personally believe, probable) that they get rid of the unauthenticated endpoints at all, which would only leave us the authorized API which is confirmed to cost money.

**Comment by sigaloid** [comment]
Okay, no backpedaling from Reddit admins at all, they've basically confirmed the change: 

https://r.nf/r/RedReader/comments/13ylk42/update_3_reddit_effectively_kills_off_third_party/

> They confirmed to me the new cost of 3rd party apps accessing the site, which is exactly what the Apollo dev revealed -- for every 50 million requests they want $12,000.

> They won't be making exceptions for free apps.

This is **very disappointing news**. :pensive: we'll have to see whether Libreddit can live with the cut down API post-change.

**Comment by TruncatedDinosour** [comment]
web scraping is not an option right ?

**Comment by sigaloid** [comment]
It's technically possible, but the roadblocks would be significant. Cloudflare bot blocking, adapting to changing HTML, etc. I think spoofing the mobile clients would be a better alternative to web scraping. I've made some serious progress with this in #819. I can't speak for what they'll do in response to these API changes - maybe they severely rate-limit or even block non-authenticated mobile clients - but I can only do so much in anticipation before these changes happen. I don't _think_ they'll do this to the official client - this is the "official" (more importantly, "ad-delivering") method of viewing content, whereas the anonymous endpoints or even a regular generated OAuth token doesn't necessarily display ads.

**Comment by TruncatedDinosour** [comment]
> It's technically possible, but the roadblocks would be significant. Cloudflare bot blocking, adapting to changing HTML, etc. I think spoofing the mobile clients would be a better alternative to web scraping. I've made some serious progress with this in #819. I can't speak for what they'll do in response to these API changes - maybe they severely rate-limit or even block non-authenticated mobile clients - but I can only do so much in anticipation before these changes happen. I don't _think_ they'll do this to the official client - this is the "official" (more importantly, "ad-delivering") method of viewing content, whereas the anonymous endpoints or even a regular generated OAuth token doesn't necessarily display ads.

thank you for the response, but yh, i rlly cant think a better idea than that, yh, good luck on this :)

**Comment by sigaloid** [comment]
As per [spez's AMA:](https://www.reddit.com/r/reddit/comments/145bram/addressing_the_community_about_changes_to_our_api/)

> Effective July 1, 2023, the rate limits to use the Data API free of charge are:
> 100 queries per minute per OAuth client id if you are using OAuth authentication and 10 queries per minute if you are not using OAuth authentication.
> Today, over 90% of apps fall into this category and can continue to access the Data API for free.

This is a pretty big change - 100req/min for free.
 
Some things to think about:

* Will libreddit fit under these limits, or will we need any sort of bypass/spoofing to continue getting access? How many requests per second do the instances get on average? If the reddit requests are under 100/min, can we get away with just using a regular OAuth token under regular free limits?
	* I want to push out 0.31.0 as soon as possible to be able to see data on this on the stats page. I currently have no idea of what traffic instances get. #808 
* Are my changes in #819 going to apply under these limits? I doubt they will consider the official app under these rate limits, but it's a possibility as others have mentioned. If it is, will we be able to fit under these limits? Can we change to an infrastructure where we can distribute the load over multiple tokens? If they don't apply these limits to the official app, how can we best match behavior to spoof an official client?

**Comment by r7l** [comment]
I guess it would be allot of work but maybe it would be possible to implement some sort of internal API network between instances in order to spread requests through any of the known instances. A single pageload by one user could the be randomly passed through any other instance. This might prevent hitting limits on popular instances and spare users from instance hopping.

**Comment by flyingfcuk** [comment]
We can piggyback on RedReader's OAth, as spez said it will be unlimited for them.
https://old.reddit.com/r/RedReader/comments/145du4j/update_4_redreader_granted_noncommercial/

To summarize:
RedReader can continue to operate as a free and open source app.
There will be no ads, monetization, etc.

They won't be rate-limiting RedReader at least for the first month, then a better solution could be found.

**Comment by sigaloid** [comment]
That's only if we retrieve the OAuth logins from the app via reverse engineering. If we do that, and cause a huge traffic boost under RedReader's name, it might lose them their exempt status. I'd like to avoid that if I can - there's no reason to paint a target on RedReader's back unnecessarily (an independent, non-commercial app), especially since the equivalent can be done with the official app without the same risks.

**Comment by flyingfcuk** [comment]
I agree it's not the best longterm solution, but the userbase is huge, 100K+ downloads only from Play Store, and probably around the same number from F-Droid and Github combined. This is also a "full client", supports accounts and posting, I assume it generates millions of requests per day, so the load will be quite negligible for something like Libreddit with a few instances. Getting the key from there is easy, just need to run it on a rooted device with self signed root CA using something like BurpSuite, then examine the requests to Reddit's API.

**Comment by erlend-sh** [comment]
Any chance libreddit could be adapted to support the Lemmy API as an alternative to Reddit?

https://users.rust-lang.org/t/building-a-better-r-rust-together/95194/

**Comment by a7maadf** [comment]
> This is a pretty big change - 100req/min for free.

I guess in that case, only private libreddit instances are going to survive

**Comment by sigaloid** [comment]
It depends - Libreddit has some decent caching enabled so depending on the level of traffic, smaller instances might survive. Also if we set up some sort of auto-balancer.

**Comment by a7maadf** [comment]
Oh wow! Thanks you for the efforts you've put towards this great project.

**Comment by PiotrKrzyzek** [comment]
> For those of us that like to keep our instance public but only share with friends / family could adding forward auth be reasonable or would you just suggest to use something like basic auth on our reverse proxies for libreddit instead?

I'm planning on putting it behind Authentik where I can then control users and they can log in via Oauth. Makes it easier for non-tech friends and family. :)

**Comment by r7l** [comment]
What's the latest news on the subject? Libreddit is still working as usual for me. Wasn't it 2 days ago when the API would be shut down for public? Sorry for the dumb question but i've not been following the drama to much since most of the subs been down anyways and i am more of a passive Libreddit user then i am an active user of Reddit.

**Comment by r7l** [comment]
Thanks. It said `These changes will be effective on June 19, 2023.` in OP.

**Comment by sigaloid** [comment]
They haven't made the changes yet, but more relevant is that we've discussed it, and we decided we can't open ourselves up to legal liability by spoofing any clients. We're going to have to provide an option for the instance operators to add their own API key. Whether or not we provide instructions on how to retrieve one, and via what avenue, is something to be seen, but we **cannot** distribute code that does this. It's not clear to what extent the API is changing, and we need to see those changes properly first before we make a change.

**Comment by r7l** [comment]
While it's understandable it defeats the purpose of using Libreddit for privacy reasons.

**Comment by sigaloid** [comment]
It reduces the privacy impact to the instance operator (hopefully divided among many, many users), just like before, since all requests come from the same IP address and as such are trivially categorized into one.

This is something the _instance operator_ will be able to add, not users, and ideally instance operators can generate these in a way that is privacy-respecting, ie, may not necessarily require a Reddit account.

**Comment by cryo420** [comment]
> An important question we need to consider is what we do if they lock down unauthenticated requests. Is webpage scraping something that we should implement? I think it's really important to have something like Libreddit but it's a different ballgame if it's a cat and mouse game rather than just piggy-backing on the API routes which is (assumed to be) allowed to be used.
> 
> [Web scraping _is_ legal, technically, in the USA.](https://techcrunch.com/2022/04/18/web-scraping-legal-court/) But it might not be good to stoop into the grey area especially in other countries.
> 
> We'll have to discuss this if it happens, but it's good to ask early so we can think about it.

this is 1 of the few eternally secure ways, so yes we should

**Comment by cryo420** [comment]
rss is possible via a wget scraper that feeds to a rss service system hosted on the same device, allowing instances & selfhosting

**Comment by mpeter50** [comment]
If I'm not mistaken and the API changes going into effect will immediately affect libreddit, it may be useful to pin this issue or an other one so that less people will start opening new issues when they notice that libreddit stopped working.

**Comment by thechesu** [comment]
Libreddit is still working for me.   Although I'm the only user.

**Comment by ashtron** [comment]
It seems many instances are still up and running (pretty much all the ones I tried). How is this possible? Am I missing something?

**Comment by sigaloid** [comment]
We aren't affected yet, for reasons unknown the changes they made haven't affected us. Likewise with teddit, etc.

**Comment by r7l** [comment]
Has anything been changed at all from Reddit? At least those 3rd party mobile apps i am using aren't affected as well.

**Comment by sigaloid** [comment]
Yes, they revoked Apollo's API key and many other third party apps no longer work, I guess the ones left have made a deal with Reddit? I can't be sure.

**Comment by PaperCutBad** [comment]
Infinity for Reddit and Stealth are still functional. Although Stealth has added option for web scraping as source along with teddit.

**Comment by Trit34** [comment]
> We aren't affected yet, for reasons unknown the changes they made haven't affected us. Likewise with teddit, etc.

Could we consider lucky, then? I hope both will continue to work.

**Comment by mpeter50** [comment]
> Yes, they revoked Apollo's API key

I thought the Apollo dev revoked his key himself, to avoid any costs that would incur by having the clients still use the API after.
I think he was talking about reddit only charging for the API usage at the end of the month or something like that.

**Comment by r7l** [comment]
Is libreddit even using a key? I've not seen it being part of the config.

**Comment by mpeter50** [comment]
According to [this](https://github.com/libreddit/libreddit/issues/785#issuecomment-1514809796), it doesn't

**Comment by sigaloid** [comment]
> I thought the Apollo dev revoked his key himself

They revoked Apollo's specifically early, according to Christian: 
![image](https://github.com/libreddit/libreddit/assets/69441971/cfb1a35b-f165-4ee2-b503-3adf02a58573)

https://twitter.com/ChristianSelig/status/1674928286781124611

**Comment by IAmChrisAMA** [comment]
Is there a reason libreddit still works? Will it persist this way or will it eventually not work?

**Comment by sigaloid** [comment]
I cannot predict whether or not they'll fix it. We're past the timeline they gave so I assume it's not like it'll stop working at x date.

**Comment by thechesu** [comment]
> Is there a reason libreddit still works? Will it persist this way or will it eventually not work?

I was told on[ Reddit](https://www.reddit.com/r/Save3rdPartyApps/comments/14n9yyk/comment/jqn5uzu/?utm_source=share&utm_medium=web2x&context=3) that Libreddit doesn't use the api.

**Comment by sigaloid** [comment]
The wording in that comment isn't *quite* correct. Yes, it's technically true that we don't use the API (I admit, I've said as much). But that doesn't mean we scrape anything. Reddit could block us without blocking Google SEO or anything (they could regardless, by looking for headers, ASN's, etc but that's neither here nor there). 

We use the anonymous JSON endpoints that have no agreement or ToS, no key, etc. The unspoken agreement **before this change** was that if you identify yourself and be a good internet neighbor (rate limits, internal caching, etc), Reddit would allow it and generally be lax with rate limits, bursts, etc. Libreddit spent a lot of time ensuring we do this, we cache responses for a 5 minute TTL, identify ourselves properly via the user-agent, etc. Whether or not this unspoken agreement will be upheld in the future, I can't say (it's at least an oversight for them to not have broken us yet). I can say that it's at least a little bit of a good thing that we haven't been hit by the changes when we were supposed to, but I can't say they won't patch it _tomorrow_. There's even less of a standard for them to hold themselves to compared to the API routes, which are an official service. 

TLDR: The anonymous endpoints are undocumented, without official mention, and with no guarantee or commitment.

**Comment by IAmChrisAMA** [comment]
> > Is there a reason libreddit still works? Will it persist this way or will it eventually not work?
> 
> Well, don’t complain about it, then! And as long as they don’t broke anything again on their side, I think that it will continue to work.

I'm not complaining (not sure if that vibe was being broadcasted) far from it.
